### PR TITLE
fix a race condition in handleChangeAssignment

### PR DESCRIPTION
**Pull request by Jyouping**
As handleChangeAssignment currently uses a thread pool, there is a race condition that
the previous assignment may actually process later than the new assignment. This will make
the follower instances don't contain the latest assignment from Zookeeper. The lack of synchronization in handleAssignmentChange may also end up creating unmonitored consumer, causing a rebalance issue. This fix set the thread pool size to 1 as default to avoid above issues.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/CoordinatorConfig.java [hunk 1]
```diff
@@ -51,6 +52,7 @@ public CoordinatorConfig(Properties config) {
     _retryIntervalMs = _properties.getInt(CONFIG_RETRY_INTERVAL, 1000 /* 1 second */);
     _heartbeatPeriodMs = _properties.getLong(CONFIG_HEARTBEAT_PERIOD_MS, Duration.ofMinutes(1).toMillis());
     _defaultTransportProviderName = _properties.getString(CONFIG_DEFAULT_TRANSPORT_PROVIDER, "");
+    _assignmentChangeThreadPoolThreadCount = _properties.getInt(CONFIG_DEFAULT_ASSIGNMENT_THREAD_COUNT, 1);
```
Comment (gauravrkulkarni) [inline datastream-server/src/main/java/com/linkedin/datastream/server/CoordinatorConfig.java]: If making the count more than 1 is going to cause an issue, do you think we should just skip reading the value from config? Might want to ignore the config value and deprecate the config?

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java [hunk 2]
```diff
@@ -119,7 +119,7 @@ public AbstractKafkaConnector(String connectorName, Properties config, GroupIdCo
   protected abstract AbstractKafkaBasedConnectorTask createKafkaBasedConnectorTask(DatastreamTask task);
 
   @Override
-  public void onAssignmentChange(List<DatastreamTask> tasks) {
+  public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
```
Comment (gauravrkulkarni) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java]: do all connectors need this to be synchronized?

File: datastream-server/src/main/java/com/linkedin/datastream/server/CoordinatorConfig.java (hunk 1 above)
Comment (Jyouping) [reply datastream-server/src/main/java/com/linkedin/datastream/server/CoordinatorConfig.java]: I'm fine with that. I think the original design is to have one thread to per connector, but it requires more change to the thread pool, might be better to fix it into 1.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java (hunk 2 above)
Comment (Jyouping) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java]: Technically, it depends on connector implement this, a connector can choose to implement a fine grained logic themselves. I also see other connectors, like a fileConnectopr implement synchronized on the entire method though.

Comment (gauravrkulkarni) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java]: Cool! Might want to consider making it a part of the documentation.

File: datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java [hunk 3]
```diff
@@ -271,6 +266,9 @@ public void start() {
       ConnectorInfo connectorInfo = _connectors.get(connectorType);
       ConnectorWrapper connector = connectorInfo.getConnector();
 
+      // Creating a separate thread pool for making the onAssignmentChange calls to the connector
+      _assignmentChangeThreadPool.put(connectorType, Executors.newSingleThreadExecutor());
```
Comment (gauravrkulkarni) [inline datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java]: nice!

**Pull request merged**